File: FAO_case_study/algorithms/input/reader.py
import networkx as nx
from typing import List, Dict
from .task import Task
from .robot import Robot


class Reader:
    def read_file_to_graph(self, graph_file: str) -> nx.Graph:
        """Read graph from file and return a NetworkX weighted undirected graph."""
        arc_graph = nx.Graph()

        with open(graph_file, 'r') as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                parts = line.split()
                a = int(parts[0])
                b = int(parts[1])
                c = float(parts[2])

                arc_graph.add_node(a)
                arc_graph.add_node(b)
                arc_graph.add_edge(a, b, weight=c)

        return arc_graph

    def read_file_to_tasks(self, tasks_file: str) -> List[Task]:
        """Read tasks from file and return a list of Task objects."""
        tasks = []

        with open(tasks_file, 'r') as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                parts = line.split()

                # Skip header line (task count) - has only 1 element
                if len(parts) == 1:
                    continue

                task = Task(
                    task_id=int(parts[0]),
                    size=float(parts[1]),
                    arrive_time=int(parts[2])
                )
                tasks.append(task)

        return tasks

    def read_file_to_robots(self, robots_file: str) -> List[Robot]:
        """Read robots from file and return a list of Robot objects."""
        robots = []

        with open(robots_file, 'r') as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                parts = line.split()
                robot = Robot(
                    robot_id=int(parts[0]),
                    capacity=float(parts[1]),
                    load=0.0,
                    tasks_list=[],
                    group_id=int(parts[2])
                )
                robots.append(robot)

        return robots

File: cascadingFailuresTaskMigration/src/main/java/input/Reader.java
package input;

import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Reader {
    public DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> readFileToGraph(String graphFile) throws IOException {
        BufferedReader bufferedReader;
        String cur[];
        bufferedReader = new BufferedReader(new FileReader(graphFile));
        String line = null;
        //用JGraph定义一个无向的权重图
        DefaultUndirectedWeightedGraph<Integer,DefaultWeightedEdge> arcGraph =
                new DefaultUndirectedWeightedGraph<>(DefaultWeightedEdge.class);
        int a, b;
        double c ;
        while ((line = bufferedReader.readLine()) != null) {
            cur = line.split(" ");
            //1.向图中添加点
            a = Integer.parseInt(cur[0]);
            b = Integer.parseInt(cur[1]);
            c = Integer.parseInt(cur[2]);
            arcGraph.addVertex(a);
            arcGraph.addVertex(b);

            //2.向图中添加边

            DefaultWeightedEdge defaultWeightedEdge = arcGraph.addEdge(a,b);
            arcGraph.setEdgeWeight(defaultWeightedEdge,c);
        }

        return arcGraph ;
    }
    public List<Task> readFileToTasks(String tasksFile) throws IOException {
        BufferedReader bufferedReader;
        String cur[];
        bufferedReader = new BufferedReader(new FileReader(tasksFile));
        String line = null;
        List<Task> tasks = new ArrayList<>();
        while ((line = bufferedReader.readLine()) != null) {
            cur = line.split(" ");
            Task task = new Task();
            task.setTaskId(Integer.parseInt(cur[0]));
            task.setSize(Integer.parseInt(cur[1]));
            task.setArriveTime(Integer.parseInt(cur[2]));
            tasks.add(task);
        }
        //用一个三元组来刻画tasks，TaskId(task[0])表示第i个任务的编号,Size(task[1])表示任务的大小，
        // ArriveTime(task[3])表示任务的到来时间
        return tasks;
    }

    public List<Robot> readFileToRobots(String tasksFile) throws IOException {
        BufferedReader bufferedReader;
        String cur[];
        bufferedReader = new BufferedReader(new FileReader(tasksFile));
        String line = null;
        List<Robot> robots = new ArrayList<>();
        while ((line = bufferedReader.readLine()) != null) {
            cur = line.split(" ");
            Robot robot = new Robot();
            robot.setRobotId(Integer.parseInt(cur[0]));
            robot.setCapacity(Integer.parseInt(cur[1]));
            robot.setLoad(0);
            robot.setTasksList(new ArrayList<>());
            robot.setGroupId(Integer.parseInt(cur[2]));
            robots.add(robot);
        }
        //用一个三元组来刻画robots，robot[0]表示机器人编号，robot[1]表示机器人能力，
        //robot[2]表示机器人所在group的编号
        return robots;
    }
}
